pkg/segment/segment.go

```diff
@@ -0,0 +1,153 @@
+package segment
+
+import (
+	"errors"
+	"fmt"
+	"github.com/openshift/odo/pkg/preference"
+	"github.com/openshift/odo/pkg/version"
+	"github.com/pborman/uuid"
+	"golang.org/x/term"
+	"gopkg.in/segmentio/analytics-go.v3"
+	"io/ioutil"
+	"net"
+	"os"
+	"os/user"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+// Writekey will be the API key used to send data to the correct source on Segment
+var WriteKey = "4xGV1HV7K2FtUWaoAozSBD7SNCBCJ65U"
+
+type Client struct {
+	segmentClient     analytics.Client
+	config            *preference.PreferenceInfo
+	telemetryFilePath string
+}
+
+func NewClient(config *preference.PreferenceInfo) (*Client, error) {
+	homeDir, _ := os.UserHomeDir()
+	return newCustomClient(config,
+		filepath.Join(homeDir, ".redhat", "anonymousId"),
+		analytics.DefaultEndpoint,
+	)
+}
+
+func newCustomClient(config *preference.PreferenceInfo, telemetryFilePath string, segmentEndpoint string) (*Client, error) {
+	client, err := analytics.NewWithConfig(WriteKey, analytics.Config{
+		Endpoint: segmentEndpoint,
+		DefaultContext: &analytics.Context{
+			IP: net.IPv4(0, 0, 0, 0),
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &Client{
+		segmentClient:     client,
+		config:            config,
+		telemetryFilePath: telemetryFilePath,
+	}, nil
+}
+
+// Close client connection and send the data
+func (c *Client) Close() error {
+	return c.segmentClient.Close()
+}
+
+// Upload prepares the data to be sent to segment and send it once the client connection closes
+func (c *Client) Upload(action string, duration time.Duration, err error) error {
+	// if the user has not consented for telemetry, return
+	if !c.config.GetConsentTelemetry() {
+		return nil
+	}
+
+	// obtain the anonymous ID
+	anonymousID, uerr := getUserIdentity(c.telemetryFilePath)
+	if uerr != nil {
+		return uerr
+	}
+
+	// queue the data that helps identify the user on segment
+	if err := c.segmentClient.Enqueue(analytics.Identify{
+		AnonymousId: anonymousID,
+		Traits:      addConfigTraits(c.config, traits()),
+	}); err != nil {
+		return err
+	}
+
+	// add information to the data
+	properties := analytics.NewProperties()
+	// TODO: add other properties when required
```

Comment:
TODO?

Reply:
That TODO will come into action if we decide to add more properties like I mentioned here - https://github.com/openshift/odo/pull/4482#discussion_r596599713